// @sap-ux/preview-middleware/dist/client/adp/init-dialogs.js
"use strict";

sap.ui.define(["sap/ui/core/Fragment", "sap/ui/fl/Utils", "./controllers/AddFragment.controller", "./controllers/ControllerExtension.controller", "./controllers/ExtensionPoint.controller"], function (Fragment, FlUtils, __AddFragment, __ControllerExtension, __ExtensionPoint) {
  "use strict";

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule && typeof obj.default !== "undefined" ? obj.default : obj;
  }
  const AddFragment = _interopRequireDefault(__AddFragment);
  const ControllerExtension = _interopRequireDefault(__ControllerExtension);
  const ExtensionPoint = _interopRequireDefault(__ExtensionPoint);
  var DialogNames = /*#__PURE__*/function (DialogNames) {
    DialogNames["ADD_FRAGMENT"] = "AddFragment";
    DialogNames["CONTROLLER_EXTENSION"] = "ControllerExtension";
    DialogNames["ADD_FRAGMENT_AT_EXTENSION_POINT"] = "ExtensionPoint";
    return DialogNames;
  }(DialogNames || {});
  /**
  * Handler for enablement of Extend With Controller context menu entry
  *
  * @param overlays Control overlays
  * @param syncViewsIds Runtime Authoring
  *
  * @returns boolean
  */
  const isControllerExtensionEnabled = (overlays, syncViewsIds) => {
    const clickedControlId = FlUtils.getViewForControl(overlays[0].getElement()).getId();
    return overlays.length <= 1 && !syncViewsIds.includes(clickedControlId);
  };

  /**
   * Determines whether the fragment command should be enabled based on the provided overlays.
   *
   * @param {ElementOverlay[]} overlays - An array of ElementOverlay objects representing the UI overlays.
   * @returns {boolean} True if the fragment command is enabled, false otherwise.
   */
  const isFragmentCommandEnabled = overlays => {
    if (overlays.length === 0) {
      return false;
    }
    const control = overlays[0].getElement();
    const hasStableId = FlUtils.checkControlId(control);
    return hasStableId && overlays.length <= 1;
  };

  /**
   * Determines the text that should be displayed for the Add Fragment context menu item.
   *
   * @param {ElementOverlay} overlay - An ElementOverlay object representing the UI overlay.
   * @returns {string} The text of the Add Fragment context menu item.
   */
  const getAddFragmentItemText = overlay => {
    if (!isFragmentCommandEnabled([overlay])) {
      return 'Add: Fragment (Unavailable due to unstable ID of the control or its parent control)';
    }
    return 'Add: Fragment';
  };

  /**
   * Handler for new context menu entry
   *
   * @param overlay Control overlays
   * @param rta Runtime Authoring
   * @param dialogName Dialog name
   * @param extensionPointData Control ID
   */
  async function handler(overlay, rta, dialogName, extensionPointData) {
    let controller;
    switch (dialogName) {
      case DialogNames.ADD_FRAGMENT:
        controller = new AddFragment(`open.ux.preview.client.adp.controllers.${dialogName}`, overlay, rta);
        break;
      case DialogNames.CONTROLLER_EXTENSION:
        controller = new ControllerExtension(`open.ux.preview.client.adp.controllers.${dialogName}`, overlay, rta);
        break;
      case DialogNames.ADD_FRAGMENT_AT_EXTENSION_POINT:
        controller = new ExtensionPoint(`open.ux.preview.client.adp.controllers.${dialogName}`, overlay, rta, extensionPointData);
        break;
    }
    const id = dialogName === DialogNames.ADD_FRAGMENT_AT_EXTENSION_POINT ? `dialog--${dialogName}` : undefined;
    const dialog = await Fragment.load({
      name: `open.ux.preview.client.adp.ui.${dialogName}`,
      controller,
      id
    });
    await controller.setup(dialog);
  }

  /**
   * Adds a new item to the context menu
   *
   * @param rta Runtime Authoring
   * @param syncViewsIds Ids of all application sync views
   */
  const initDialogs = (rta, syncViewsIds) => {
    const contextMenu = rta.getDefaultPlugins().contextMenu;
    contextMenu.addMenuItem({
      id: 'ADD_FRAGMENT',
      text: getAddFragmentItemText,
      handler: async overlays => await handler(overlays[0], rta, DialogNames.ADD_FRAGMENT),
      enabled: isFragmentCommandEnabled,
      icon: 'sap-icon://attachment-html'
    });
    contextMenu.addMenuItem({
      id: 'EXTEND_CONTROLLER',
      text: 'Extend With Controller',
      handler: async overlays => await handler(overlays[0], rta, DialogNames.CONTROLLER_EXTENSION),
      icon: 'sap-icon://create-form',
      enabled: overlays => isControllerExtensionEnabled(overlays, syncViewsIds)
    });
  };
  var __exports = {
    __esModule: true
  };
  __exports.DialogNames = DialogNames;
  __exports.isControllerExtensionEnabled = isControllerExtensionEnabled;
  __exports.isFragmentCommandEnabled = isFragmentCommandEnabled;
  __exports.getAddFragmentItemText = getAddFragmentItemText;
  __exports.handler = handler;
  __exports.initDialogs = initDialogs;
  return __exports;
});
